solutions/USACO jan16-radio-contact.mdx

```diff
@@ -0,0 +1,128 @@
+---
+id: usaco-598
+source: USACO Gold 2016 January
+title: Radio Contact
+author: Neo Wang
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_radio_gold_jan16.html)
+
+## Solution
+
+The key $\texttt{dp}$ observation to make here is already stated in the problem statement: "Farmer John can either stay put at his current location, or take one step forward", and likewise for Bessie. Thus, when constructing our $\texttt{dp}$, we can set $\texttt{dp}[i][j]$ equal to the best distance at Farmer John's move $i$ and Bessie's move $j$. 
+
+Notice that it is hard to calculate a cumulative distance if we leave the string unprocessed (meaning that we read from the string directly). To resolve this, we can simply calculate the coordinates $(i,j)$ at every step of Bessie and Farmer John's path. In our implementation, we map each character to their appropriate `dx`, `dy` arrays in order to apply the appropriate changes, storing Farmer John's position at move $i$ as $\texttt{jl}[i]$ and Bessie's position at move $j$ at $\texttt{bl}[j]$.
+
+After this is done, we start building our $\texttt{dp}$:
+
+Either one of the following happens:
+
+1. Farmer John takes a step $(i+1)$: $\texttt{dp}[i+1][j] = \min(\texttt{dp}[i+1][j], \texttt{dp}[i][j] + \text{dist}(\texttt{jl}[i+1], \texttt{bl}[j]))$
+
+2. Bessie takes a step $(j+1)$: $\texttt{dp}[i][j+1] = \min(\texttt{dp}[i][j+1], \texttt{dp}[i][j] + \text{dist}(\texttt{jl}[i], \texttt{bl}[j+1]))$
+
+3. Both Farmer John and Bessie take steps $(i+1)$ and $(j+1)$: $\texttt{dp}[i+1][j+1] = \min(\texttt{dp}[i+1][j+1], \texttt{dp}[i][j] + \text{dist}(\texttt{jl}[i+1], \texttt{bl}[j+1]))$
+
+<LanguageSection>
+
+<CPPSection>
+
+## Implementation
+
+```cpp
+#include <bits/stdc++.h> // see /general/running-code-locally
+using namespace std;
+
+using ll = long long;
+
+using vi = vector<int>;
+#define pb push_back
+#define all(x) begin(x), end(x)
+#define sz(x) (int)(x).size()
+
+using pi = pair<int,int>;
+#define f first
+#define s second
+#define mp make_pair
+
+void setIO(string name = "") {
+	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
+	if (sz(name)) {
+		freopen((name+".in").c_str(), "r", stdin); // see /general/io
+		freopen((name+".out").c_str(), "w", stdout);
+	}
+}
+
+int N, M;
+
+const int INF = 1e9+7;
+const int MX = 1e3+1;
+
+int sq(int a) {
+	return a * a;
+}
+
+int dist(pi a, pi b) { // distance squared between two points
+	return sq(a.f - b.f) + sq(a.s - b.s);
+}
+
+map<char, int> md{{'N', 0}, {'E', 1}, {'S', 2}, {'W', 3}};
+
+const int dx[4] = {0, 1, 0, -1};
+const int dy[4] = {1, 0, -1, 0};
+
+// i = current step of John
+// j = current step of Bessie
+
+int dp[MX][MX];
+
+int main() {
+	setIO("radio");
+
+	cin >> N >> M;
+	
+	fill_n(dp[0], MX*MX, INF);
```

Comment:
I think line 84 should not go in between the reading of `N` and `M` and the initialization of the vectors.